lightning/src/ln/onion_utils.rs

```diff
@@ -404,24 +411,31 @@ pub(super) fn process_onion_failure<T: secp256k1::Signing, L: Deref>(secp_ctx: &
 										let is_chan_update_invalid = match error_code & 0xff {
 											7 => false,
 											11 => amt_to_forward > chan_update.contents.htlc_minimum_msat,
-											12 => {
-												let new_fee = amt_to_forward.checked_mul(chan_update.contents.fee_proportional_millionths as u64).and_then(|prop_fee| { (prop_fee / 1000000).checked_add(chan_update.contents.fee_base_msat as u64) });
-												new_fee.is_some() && route_hop.fee_msat >= new_fee.unwrap()
-											}
+											12 => amt_to_forward
+												.checked_mul(chan_update.contents.fee_proportional_millionths as u64)
+												.map(|prop_fee| prop_fee / 1_000_000)
+												.and_then(|prop_fee| prop_fee.checked_add(chan_update.contents.fee_base_msat as u64))
+												.map(|fee_msats| route_hop.fee_msat >= fee_msats)
+												.unwrap_or(false),
 											13 => route_hop.cltv_expiry_delta as u16 >= chan_update.contents.cltv_expiry_delta,
 											14 => false, // expiry_too_soon; always valid?
 											20 => chan_update.contents.flags & 2 == 0,
 											_ => false, // unknown error code; take channel_update as valid
 										};
-										network_update = if is_chan_update_invalid {
+										if is_chan_update_invalid {
 											// This probably indicates the node which forwarded
 											// to the node in question corrupted something.
-											Some(NetworkUpdate::ChannelClosed {
+											network_update = Some(NetworkUpdate::ChannelClosed {
 												short_channel_id: route_hop.short_channel_id,
 												is_permanent: true,
-											})
+											});
 										} else {
-											Some(NetworkUpdate::ChannelUpdateMessage {
+											// Make sure the ChannelUpdate contains the expected
+											// short channel id.
+											if failing_route_hop.short_channel_id == chan_update.contents.short_channel_id {
+												short_channel_id = Some(failing_route_hop.short_channel_id);
+											}
+											network_update = Some(NetworkUpdate::ChannelUpdateMessage {
```

Comment:
Wait, hmm, so if `failing_route_hop.short_channel_id != chan_update.contents.short_channel_id` here we'll return a `ChannelUpdate` with a SCID unrelated to the node that sent us the update (presumably then ignoring the update because its invalid/signatures dont match) and then have no SCID to blame for the failure, so we'll just retry the payment again over the same path?

Reply:
We'll hit `short_channel_id = Some(route_hop.short_channel_id)` on line 454 and not route through that hop, right?